File: backend/product-service/product-service/src/main/java/com/shopsphere/product_service/service/ProductService.java
package com.shopsphere.product_service.service;

import com.shopsphere.product_service.dto.ProductRequest;
import com.shopsphere.product_service.entity.Category;
import com.shopsphere.product_service.entity.Product;
import com.shopsphere.product_service.repository.CategoryRepository;
import com.shopsphere.product_service.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;

    @CacheEvict(value = "products", allEntries = true)
    public Product create(ProductRequest request) {
        Category category = categoryRepository.findById(request.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found with id: " + request.getCategoryId()));

        Product product = Product.builder()
                .name(request.getName())
                .description(request.getDescription())
                .price(request.getPrice())
                .stockQuantity(request.getStockQuantity())
                .imageUrl(request.getImageUrl())
                .active(true)
                .category(category)
                .build();
        return productRepository.save(product);
    }

    @Cacheable(value = "products", key = "'all'")
    public List<Product> findAll() {
        return productRepository.findAll();
    }

    @Cacheable(value = "product", key = "#id")
    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));
    }

    public List<Product> findByCategory(Long categoryId) {
        return productRepository.findByCategoryId(categoryId);
    }

    public List<Product> searchByName(String name) {
        return productRepository.findByNameContainingIgnoreCase(name);
    }

    public List<Product> filterByPrice(BigDecimal minPrice, BigDecimal maxPrice) {
        return productRepository.findByPriceBetween(minPrice, maxPrice);
    }

    public Page<Product> findAllPaginated(int page, int size, String sortBy, String direction) {
        Sort sort = direction.equalsIgnoreCase("desc")
                ? Sort.by(sortBy).descending()
                : Sort.by(sortBy).ascending();
        return productRepository.findAll(PageRequest.of(page, size, sort));
    }

    @Caching(evict = {
            @CacheEvict(value = "products", allEntries = true),
            @CacheEvict(value = "product", key = "#id")
    })
    public Product update(Long id, ProductRequest request) {
        Category category = categoryRepository.findById(request.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found with id: " + request.getCategoryId()));

        Product product = findById(id);
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setStockQuantity(request.getStockQuantity());
        product.setImageUrl(request.getImageUrl());
        product.setCategory(category);
        return productRepository.save(product);
    }

    @Caching(evict = {
            @CacheEvict(value = "products", allEntries = true),
            @CacheEvict(value = "product", key = "#id")
    })
    public Product updateStock(Long id, Integer quantity) {
        Product product = findById(id);
        product.setStockQuantity(quantity);
        return productRepository.save(product);
    }

    @Caching(evict = {
            @CacheEvict(value = "products", allEntries = true),
            @CacheEvict(value = "product", key = "#id")
    })
    public Product decrementStock(Long id, Integer quantity) {
        Product product = findById(id);
        int current = product.getStockQuantity();
        if (current < quantity) {
            throw new RuntimeException("Insufficient stock for product " + id
                    + " | Available: " + current + " | Requested: " + quantity);
        }
        product.setStockQuantity(current - quantity);
        return productRepository.save(product);
    }

    @Caching(evict = {
            @CacheEvict(value = "products", allEntries = true),
            @CacheEvict(value = "product", key = "#id")
    })
    public Product increaseStock(Long id, Integer quantity) {
        Product product = findById(id);
        product.setStockQuantity(product.getStockQuantity() + quantity);
        return productRepository.save(product);
    }

    @Caching(evict = {
            @CacheEvict(value = "products", allEntries = true),
            @CacheEvict(value = "product", key = "#id")
    })
    public void delete(Long id) {
        Product product = findById(id);
        productRepository.delete(product);
    }
}

File: frontend/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Product, ProductCategory, SpringPage } from '../models/models';

@Injectable({ providedIn: 'root' })
export class ProductService {
  constructor(private readonly http: HttpClient) {}

  findAll(): Observable<Product[]> {
    return this.http.get<Product[]>(`${environment.apiUrl}/api/products`);
  }

  getCategories(): Observable<ProductCategory[]> {
    return this.http.get<ProductCategory[]>(`${environment.apiUrl}/api/categories`);
  }

  findPaginated(page: number, size: number, sortBy = 'id', direction = 'asc'): Observable<SpringPage<Product>> {
    return this.http.get<SpringPage<Product>>(
      `${environment.apiUrl}/api/products/paginated?page=${page}&size=${size}&sortBy=${sortBy}&direction=${direction}`,
    );
  }

  findById(id: number): Observable<Product> {
    return this.http.get<Product>(`${environment.apiUrl}/api/products/${id}`);
  }

  searchByName(name: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${environment.apiUrl}/api/products/search?name=${encodeURIComponent(name)}`);
  }

  findByCategory(categoryId: number): Observable<Product[]> {
    return this.http.get<Product[]>(`${environment.apiUrl}/api/products/category/${categoryId}`);
  }

  filterByPrice(minPrice: number, maxPrice: number): Observable<Product[]> {
    return this.http.get<Product[]>(
      `${environment.apiUrl}/api/products/filter/price?minPrice=${minPrice}&maxPrice=${maxPrice}`,
    );
  }

  create(request: ProductRequest): Observable<Product> {
    return this.http.post<Product>(`${environment.apiUrl}/api/products`, request);
  }

  update(id: number, request: ProductRequest): Observable<Product> {
    return this.http.put<Product>(`${environment.apiUrl}/api/products/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${environment.apiUrl}/api/products/${id}`);
  }

  updateStock(id: number, quantity: number): Observable<Product> {
    return this.http.patch<Product>(`${environment.apiUrl}/api/products/${id}/stock`, { quantity });
  }

  createCategory(request: CategoryRequest): Observable<ProductCategory> {
    return this.http.post<ProductCategory>(`${environment.apiUrl}/api/categories`, request);
  }

  updateCategory(id: number, request: CategoryRequest): Observable<ProductCategory> {
    return this.http.put<ProductCategory>(`${environment.apiUrl}/api/categories/${id}`, request);
  }

  deleteCategory(id: number): Observable<void> {
    return this.http.delete<void>(`${environment.apiUrl}/api/categories/${id}`);
  }
}

export interface ProductRequest {
  name: string;
  description: string;
  price: number;
  stockQuantity: number;
  imageUrl: string;
  categoryId: number;
}

export interface CategoryRequest {
  name: string;
  description: string;
  imageUrl: string;
}